deepchecks/utils/abstracts/weak_segment_abstract.py

```diff
@@ -213,24 +213,48 @@ def _weak_segments_search(self, data: pd.DataFrame, score_per_sample: pd.Series,
                                                              tuple(filters[feature2]), data_size,
                                                              list(data_of_segment.index)]
 
-        # Sort and drop relevant columns
-        weak_segments = weak_segments.sort_values(score_title).reset_index(drop=True)
-        if multiple_segments_per_feature:
-            result = weak_segments.drop(columns='Samples in Segment').drop_duplicates()
-            result['Samples in Segment'] = weak_segments.loc[result.index, 'Samples in Segment']
-        else:
-            used_features = set()
-            result = pd.DataFrame(columns=weak_segments.columns)
-            for _, row in weak_segments.iterrows():
-                if row['Feature1'] in used_features or row['Feature2'] in used_features:
-                    continue
-
-                result.loc[len(result)] = row
-                used_features.add(row['Feature1'])
-                if row['Feature2'] != '':
-                    used_features.add(row['Feature2'])
-
-        return result
+        # Filter and adapt the weak segments results
+        result = pd.DataFrame(columns=weak_segments.columns)
+        used_features = set()
+        for _, row in weak_segments.sort_values(score_title).iterrows():
+            new_row = row.copy()
+            if not multiple_segments_per_feature and \
+                    (row['Feature1'] in used_features or row['Feature2'] in used_features):
+                continue
+
+            # Make sure segments based on categorical features are based only on a single category
+            if row['Feature1'] in self.encoder_mapping:
+                unique_values_in_range = [x for x in self.encoder_mapping[row['Feature1']]['encoded_value'].values if
```

Comment:
issue with this logic is that it can leave us with weak segments that are not actually weak - you are checking that they are big enough (which is good), but there is no guarantee that the segment has a score than is lower than the rest of the data by more than a small margin.